### ISSUE TITLE
The locale resolution is losing the country code on iOS

### ISSUE THREAD

**Issue by saintmac-google** [flutter/flutter#128424]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

In https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/platform_view_ios.mm#L244
the function is converting a locale id into a set of language/script/country strings, but the countryCode accessor is deprecated and returns null.
https://developer.apple.com/documentation/foundation/nslocale/1643060-countrycode?language=objc
What used to be the country code is now accessible through variantCode.
 
Repro steps:
1 Ensure you have an english language listed in your language settings on your iOS device
2. call the following function
WidgetsBinding.instance.platformDispatcher
      .computePlatformResolvedLocale(<Locale>[Locale('en', 'US')]);

### Expected results

The function returns Locale('en', 'US') (language="en", countryCode="US")

### Actual results

The function returns Locale('en'). The country code has been lost.

### Code sample

<details><summary>Suggested fix for https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/platform_view_ios.mm#L244 </summary>

```objC
    NSString* countryCode = [locale countryCode];
    // countryCode has been deprecated since ios 17 and returns null.
    // Use the variantCode instead.
    if (countryCode == nullptr) {
      NSString* variantCode = [locale variantCode];
      out->emplace_back(variantCode == nullptr ? "" : variantCode.UTF8String);
    } else {
      out->emplace_back(countryCode.UTF8String);
    }
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

**Comment by xster** [comment]
P2 since Google team can't test or release

**Comment by xster** [comment]
Upstream b/282082089

**Comment by chinmaygarde** [comment]
I don't believe country code and variant code are the same thing and doubt the accuracy of the suggested fix.

The deprecation is also a part of the iOS 17 Beta which hasn't been released it. Not sure this is a P2.

Is this an instance of the application working around a deprecation by testing on iOS 17 Beta opted in devices, or, just the application needing another accessor (like `regionCode` that @dnfield suggests).

**Comment by dnfield** [comment]
I think the right thing to do here is to implement handling for `regionCode` in cases where it's supported. I don't think `variantCode` is what you actually want. If it is please reach out to me and we can try to figure this out.

@cyanglaz are you able to take a look at this (or find someone else on the iOS team who can)? This will require a little bit of dancing to get the compiler to recognize the `regionCode` symbol since it's probably not in the SDK we're using on CI, but other than that it should be a pretty straightforward availability check (if iOS 17, use region code, else use countryCode).

**Comment by cyanglaz** [comment]
Not sure if "This method is equivalent to calling the [displayNameForKey:value:](https://developer.apple.com/documentation/foundation/nslocale/1415931-displaynameforkey?language=objc) method passing the [NSLocaleCountryCode](https://developer.apple.com/documentation/foundation/nslocalecountrycode?language=objc) key and countryCode value." works, if it also returns null, then `regionCode` is probably the right fix on iOS 17+. I can work on it.

**Comment by cyanglaz** [comment]
I tested it on simulators, both countryCode and regionCode returns nil. It looks like an issue with simulator, not exactly due to the deprecation. 

There some related discussion online but not exactly the same scenarios: https://stackoverflow.com/questions/15202454/nslocalecountrycode-returns-nil

I haven't been able to test it on real device but will do so shortly.

**Comment by cyanglaz** [comment]
Tested on iOS 16 devices, and it is also return null. I will try to update one of my devices to 17 and test it.

**Comment by cyanglaz** [comment]
Tested on iOS 17 devices, the value is still null. I don't think it is an issue caused by API deprecation.
@saintmac-google When it worked for you before, do you remember what iOS version it was? Was it on device or simulator.

**Comment by saintmac-google** [comment]
It never worked before.
All I did was run on iPhone 14 Pro with iOS 16 simulator.
I initially thought it was a deprecation issue because when I was trying to investigate why the country code was returning nil, I stumbled upon this indication in the apple documentation that this message was deprecated as of iOS 17. I didn't realize at that time that I was running iOS 16. Let's just forget I said that.

**Comment by cyanglaz** [comment]
@saintmac-google Is it still a block for your release?

**Comment by saintmac-google** [comment]
I'm working around it, so it's a lesser priority now
